Virtual socket IO functions

elcallio opened the pull request:
To integrate the C-ares library with [seastar](https://github.com/scylladb/seastar), we need to be able to control the way socket IO is performed, since 

1.) We optionally use our own network stack (virtio/dpdk)
2.) We do asynchronous even more asynchronous. 

This patch set adds a virtual socket function struct to c-ares, which can be attached to an individual channel using a setter function. The socket IO code is modified to use these bound functions if set (which in turn forgoes a lot of socket options etc, with the assumption that the provider deals with such matters.

bagder commented:
Thanks for this!

I'm all for this change. There's some nits with the patches (trailing whitespace, some funny indents and a C90 compliance warning) but no biggie. I would like to see at least documentation added before we can merge and I would also love to see at least a basic test case to make sure we don't break this going forward!

elcallio commented:
@bagder Thanks for the feedback. Could you point me to the new warning? I did not spot any added ones, but then I have a long history of blindness. 
I'll fix the nits, but on the documentation - do you mean additional code comments or something else? Feel free to point explicitly at places lacking/needing more. 
Tests: yes - shame on me. I'll add some.

elcallio commented:
v2: 
* Added man page for socket function setter/struct
* Added test helper + cases (specific + run several "live" tests in dual mode)
* Cleaned up line endings + indents
Sorry for the multiple git bounces. Managed to bung up my force-commit with test fix for OSX.

elcallio commented:
ping?

gjasny commented:
Hello,

(Disclaimer: I just skimmed over your patch). How do you handle select/poll/epoll/kqueue? I'm asking because resiprocate (a popular SIP stack) has an embedded copy of ares (the predecessor) of c-ares where you could register a socket callback that notified about new sockets so they could be added to a socket-wait function. See here:

https://github.com/resiprocate/resiprocate/blob/master/rutil/dns/AresDns.cxx#L232

Unfortunately c-ares lacks that functionality.

Thanks,
Gregor

bagder commented:
Can I just mention that ares did not have that ability in version 1.1.1 (which the c-ares project started from), so if resiprocate has that, it was added by someone at a later stage. I personally wouldn't mind having support for that sort of thing in c-ares, but I don't think we need to pick one or the other, I think we could support both approaches.

elcallio commented:
@gjasny: The `socket_created` callback already exists in c-ares, and polling is the clients responsibility, so using epoll etc should indeed already be quite feasible. The purpose of this patch set however is mainly to allow IO to be virtualized to a different medium than actual system sockets (in our case, seastar native IO). Or any case where the communication cannot be created purely by a socket call.

elcallio commented:
At the risk of repeating myself: ping?
@bagder: is there anything you want beyond whats in the latest iteration of the patch set?

bagder commented:
Thanks for your work and your patience with this! I took the liberty of doing some very minor edits (C90 warning and removed trailing whitespace) and merge your work.

elcallio commented:
Awesome. Thanks @bagder.

daviddrysdale commented:
@elcallio: looks like there's a windows build failure -- could you take a look?

https://ci.appveyor.com/project/daviddrysdale/c-ares/build/1.0.244

(Aside: we couldn't get AppVeyor to work with the main c-ares repo for some reason, but it works with my fork.  Just let me know if you want me to push'n'pull something to AppVeyor)

elcallio commented:
@daviddrysdale Sorry about that; Indeed the test code does not cross-compile as well as one might have hoped. There are also some pre-existing issues with cross-compile setup for tests (missing data types luckily not referenced previously) for the windows build. Now they surfaced. 
I'll send a fix-pull-request asap.
